File: 10311 Goldbach and Euler/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static int MAXN = 100000000;
    static boolean[] isComposite = new boolean[(MAXN >> 1) + 1];
    static ArrayList<Integer> primes = new ArrayList<>();
	public static void main(String[] args) throws IOException {
	    eulerSieve();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
            int n = (int)st.nval;
            if(n <= 4)
                output.append(n).append(" is not the sum of two primes!\n");
            else {
                if((n & 1) == 1) {
                    if(isComposite[n - 2 >> 1])
                        output.append(n).append(" is not the sum of two primes!\n");
                    else
                        output.append(n).append(" is the sum of 2 and ").append(n - 2).append(".\n");
                } else {
                    int temp = n - 1 >> 1;
                    if((temp & 1) == 0)
                        --temp;
                    while(temp >= 3 && (isComposite[temp >> 1] || isComposite[n - temp >> 1]))
                        temp -= 2;
                    if(temp >= 3)
                        output.append(n).append(" is the sum of ").append(temp).append(" and ").append(n - temp).append(".\n");
                    else
                        output.append(n).append(" is not the sum of two primes!\n");
                }
            }
        }
		System.out.print(output);
	}
	
	static void eulerSieve() {
        for(int n = 3; n <= MAXN; n += 2) {
            if(!isComposite[n >> 1])
                primes.add(n);
            for(int i = 0, temp; (temp = primes.get(i) * n) <= MAXN; ++i) {
                isComposite[temp >> 1] = true;
                if(n % primes.get(i) == 0)
                    break;
            }
        }
    }
}

File: 10311 Goldbach and Euler/solution.c
#include <stdio.h>
#include <stdlib.h>
#define MAXN 100000000
#define GET(n) (mark[(n) >> 5] & (1u << ((n) & 31)))
#define SET(n) (mark[(n) >> 5] |= (1u << ((n) & 31)))

unsigned int mark[(MAXN >> 6) + 1];
int *primes, size;

void eulerSieve() {
    for(int n = 3; n <= MAXN; n += 2) {
        if(!GET(n >> 1)) {
            primes = (int*)realloc(primes, (size + 1) * sizeof(int));
            primes[size++] = n;
        }
        for(int i = 0, temp; (temp = primes[i] * n) <= MAXN; ++i) {
            SET(temp >> 1);
            if(n % primes[i] == 0)
                break;
        }
    }
}

int main() {
    eulerSieve();
    int n;
    while(scanf("%d", &n) == 1) {
        if(n <= 4)
            printf("%d is not the sum of two primes!\n", n);
        else {
            if(n & 1) {
                if(GET(n - 2 >> 1))
                    printf("%d is not the sum of two primes!\n", n);
                else
                    printf("%d is the sum of 2 and %d.\n", n, n - 2);
            } else {
                int temp = n - 1 >> 1;
                temp -= !(temp & 1);
                while(temp >= 3 && (GET(temp >> 1) || GET(n - temp >> 1)))
                    temp -= 2;
                if(temp >= 3)
                    printf("%d is the sum of %d and %d.\n", n, temp, n - temp);
                else
                    printf("%d is not the sum of two primes!\n", n);
            }
        }
    }
    free(primes);
    return 0;
}
